### PR TITLE
Include KB installation date in enum_patches

### PR DESCRIPTION
**Pull request by 247arjun**
This change improves the module by also having it output when a given patch package was installed (this information can also be retrieved from the WMI query); this will provide insight into how regularly and reliably a PC (and by extension, environment) patches - for example, are they late in installing patches by months, when did they last patch etc.

Currently, the output of this module only lists the KB packages installed on a Windows PC. 

## Verification

- [x] Start `msfconsole`
- [x] Spawn a Meterpreter session against a Windows target
- [x] With the session sent to background, search for `enum_patches`
- [x] Set the `SESSION` parameter to match the session of interest
- [x] `RUN` the module
- [x] **Verify** that the output now shows the date each patch was installed.

### Output with proposed changes (from a Win 7 SP1 VM)
```
[+] KB2534111 installed on 8/31/2020
[+] KB2999226 installed on 8/31/2020
[+] KB976902 installed on 11/21/2010
```
### Output without proposed changes
```
[*] KB2534111 applied
[*] KB2999226 applied
[*] KB976902 applied
```

### CODE REVIEW

**Comment by 247arjun** [comment]
A previous PR for this change was approved by the original author of this module (@mubix ).

[Link to previous PR](https://github.com/rapid7/metasploit-framework/pull/14078)

**Comment by gwillcox-r7** [comment]
Confirmed this is the same code as the previous PR that @mubix approved. Will run a few quick checks on this and get this merged in so long.

File: modules/post/windows/gather/enum_patches.rb [hunk 1]
```diff
@@ -34,32 +34,33 @@ def initialize(info={})
 
   def run
     unless load_extapi
-      print_error 'ExtAPI failed to load'
+      print_error("ExtAPI failed to load")
```
Comment (gwillcox-r7) [inline modules/post/windows/gather/enum_patches.rb:37]: So a few comments here, but in Ruby the common style is to use single quotes when not performing string interpolation and double quotes only when your performing string interpolation. Also, if the function only has a single parameter, its common to exclude the `(` and `)` characters.

On this basis I'm not sure this change is really needed.

File: modules/post/windows/gather/enum_patches.rb [hunk 2]
```diff
@@ -34,32 +34,33 @@ def initialize(info={})
...
     if objects[:values].nil?
       kb_ids = []
     else
-      kb_ids = objects[:values].reject(&:nil?).map { |kb| kb[0] }
+      kb_ids = objects[:values].reject(&:nil?).map { |kb| kb }
     end
 
     if kb_ids.empty?
-      print_status 'Found no patches installed'
+      print_status("Found no patches installed")
```
Comment (gwillcox-r7) [inline modules/post/windows/gather/enum_patches.rb:55]: See comment above for info on why these changes aren't really needed and actually break Ruby style convention.

File: modules/post/windows/gather/enum_patches.rb [hunk 3]
```diff
@@ -34,32 +34,33 @@ def initialize(info={})
...
+      print_error("ExtAPI failed to load")
       return
     end
 
     begin
-      objects = session.extapi.wmi.query("SELECT HotFixID FROM Win32_QuickFixEngineering")
+      objects = session.extapi.wmi.query("SELECT HotFixID, InstalledOn FROM Win32_QuickFixEngineering")
     rescue RuntimeError
-      print_error "Known bug in WMI query, try migrating to another process"
+      print_error("Known bug in WMI query, try migrating to another process")
```
Comment (gwillcox-r7) [inline modules/post/windows/gather/enum_patches.rb:44]: See comment above for why one doesn't really need to change this as `(` and `)` are explicitly included when you call a function with only one parameter.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]
Some minor changes, mostly just to remove stuff that doesn't make sense from a RuboCop point of view for standards compliance. Otherwise the code looks fine 👍

File: modules/post/windows/gather/enum_patches.rb (hunk 1 above)
Comment (247arjun) [reply modules/post/windows/gather/enum_patches.rb]: Fixed in next commit.

File: modules/post/windows/gather/enum_patches.rb (hunk 2 above)
Comment (247arjun) [reply modules/post/windows/gather/enum_patches.rb]: Fixed in next commit.

File: modules/post/windows/gather/enum_patches.rb (hunk 3 above)
Comment (247arjun) [reply modules/post/windows/gather/enum_patches.rb]: I've changed this - I removed the parentheses and changed it to a single quote (per your guidelines above).

**Comment by gwillcox-r7** [comment]
New alterations look good, will merge this in once it passes the build checks 👍  Thanks for your contribution @247arjun I know the team is really looking forwards to pushing this PR into the framework, and congrats on your first Metasploit PR! 🥳

**Comment by 247arjun** [comment]
> New alterations look good, will merge this in once it passes the build checks 👍 Thanks for your contribution @247arjun I know the team is really looking forwards to pushing this PR into the framework, and congrats on your first Metasploit PR! 🥳

I am so excited! My first Metasploit PR. I am definitely going to print out this page for my own archival (as if GitHub sending code to an Arctic Vault wasn't enough)... Thank you Grant for helping me out.

**Comment by h00die** [comment]
Don't forget to update the module docs with the new output

**Comment by 247arjun** [comment]
> Don't forget to update the module docs with the new output

I've updated the description in the same script. Is there anywhere else that needs updating?

**Comment by gwillcox-r7** [comment]
@247arjun I think this also needs updating, which is what @h00die was referring to `documentation/modules/post/windows/gather/enum_patches.md`. More specifically you will want to update the output under the section titled `### Windows 7 (6.1 Build 7601, Service Pack 1).`, assuming that is the version you tested on. Otherwise, remove this section, and and a section starting with `###` followed by Windows version you tested it on.

Just need to place that inside the triple single quotes like the example shows there.

**Comment by gwillcox-r7** [comment]
On second examination this line will also need to be updated:

```
  This module will attempt to enumerate which patches are applied to a
  windows system based on the result of the WMI query: `SELECT HotFixID FROM Win32_QuickFixEngineering`.
```

To this

```
  This module will attempt to enumerate which patches are applied to a
  windows system based on the result of the WMI query: `SELECT HotFixID, InstalledOn FROM Win32_QuickFixEngineering`.
```

**Comment by gwillcox-r7** [comment]
@247arjun Going to go in and make the changes I mentioned above so long and get this merged. Sorry for the delay!

**Comment by gwillcox-r7** [comment]
@247arjun Made a ninja commit with https://github.com/rapid7/metasploit-framework/commit/20e4b3e71fa03e2857e0c875109a5bde926a448f to address the documentation issues we mentioned above.

**Comment by gwillcox-r7** [comment]
# Release Notes

This update alters the behavior of the `enum_patches.rb` module so that it not only reports what patches are installed on a Windows system, but also when they were installed, which can be useful information for pentesters who want to gather information on an organization's patch history.

**Comment by 247arjun** [comment]
Thanks so much @mubix, @gwillcox-r7 and @h00die for your help with this PR. 

Woot woot! My first submission to Metasploit! I'm so happy!

**Comment by pbarry-r7** [comment]
# Release Notes

Updated <code>post/windows/gather/enum_patches</code> to additionally report <b>when</b> patches were installed on a Windows target.
